### PR TITLE
Implement the new plugin interface infrastructure

### LINKED ISSUE #359: Implement New Plugin Interface Infrastructure
The plugin design has changed a bit over time, and we're becoming inconsistent in the plugin design. The goal of this story is to re-establish common patterns among all of the plugins, and to provide them with greater flexibility.

## AC:
- [x] A class is created for the contribution points data structure
   - Wraps a dictionary of contribution point enums keys to their corresponding contributions
   - Validates that the `AaC Commands` contribution is a list of  type `aac.cli.aac_command.AacCommand`

- [x] The plugin contributions class exposes available contribution points via an enum
  - Available Contribution Points:
    - [x] AaC Commands
    - [x] Validations
    - [x] Definitions contributed by the plugin

- [x] All plugins provide a plugin metadata class for registration and management purposes
  - The metadata class includes the following attributes:
    - [x] The plugin's name
    - [x] A contribution points data structure

- [x] Add a new pluggy hook function to register plugins via their new metadata class

- [x] `plugin_manager` has a list of **all registered plugins**

- [x] `LanguageContext` has a list of applied plugins, a subset of **all registered plugins**

- [x] Update `plugin_manager` functions to pull from pluggy hooks and corresponding _registered_ plugin contribution points

### LINKED ISSUE #393: Migrate Plugins to New Contribution System
blocked by #359 
* Focus cross-plugin changes here
* Break-out per-plugin migration into individual stories

## AC:

- [ ] Definition contributed by plugins have their plugin definition file's source uri
- https://github.com/jondavid-black/AaC/blob/a6ea1c45909ca017497c50dcf054cebc390621ee/python/src/aac/plugins/plugin_manager.py#L96-L98

- [ ] Remove the use of hookspecs and hookimpls that aren't `get_plugin()`
- [ ] `plugin_manager` only uses the `get_plugin()` hook

### PR DESCRIPTION
**Pull request by cvchaparro**
Closes #359
Closes #393

Changes:
- Add a `Plugin` metadata class that provides information about a plugin including the definitions, commands, and validator plugins contributed by said plugin.
- Update all plugins to return a `Plugin` metadata object.
- Update `plugin_manager.py` to make use of the `Plugin` metadata object.
- Update `plugin_manager.py` to dynamically import plugins.
- Make a few minor updates to the `gen-plugin` plugin to enable generating the new `get_plugin` hook for new plugins.

### CODE REVIEW

File: python/src/aac/plugins/gen_design_doc/__init__.py [hunk 1]
```diff
@@ -48,10 +52,30 @@ def get_commands() -> list[AacCommand]:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_design_doc/__init__.py:79]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/gen_gherkin_behaviors/__init__.py [hunk 2]
```diff
@@ -43,4 +47,25 @@ def get_plugin_aac_definitions() -> str:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_gherkin_behaviors/__init__.py:69]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/gen_json/__init__.py [hunk 3]
```diff
@@ -43,5 +47,25 @@ def get_plugin_aac_definitions() -> str:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_json/__init__.py:69]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/gen_plant_uml/__init__.py [hunk 4]
```diff
@@ -82,4 +86,25 @@ def get_plugin_aac_definitions() -> str:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_plant_uml/__init__.py:108]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/gen_plugin/__init__.py [hunk 5]
```diff
@@ -39,10 +43,31 @@ def get_commands() -> list[AacCommand]:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_plugin/__init__.py:71]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/gen_plugin/templates/first_party/__init__.py.jinja2 [hunk 6]
```diff
@@ -47,15 +51,34 @@ def get_commands() -> list[AacCommand]:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_plugin/templates/first_party/__init__.py.jinja2:82]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/gen_plugin/templates/third_party/__init__.py.jinja2 [hunk 7]
```diff
@@ -50,11 +55,32 @@ def get_commands() -> list[AacCommand]:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_plugin/templates/third_party/__init__.py.jinja2:83]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/gen_protobuf/__init__.py [hunk 8]
```diff
@@ -43,10 +47,30 @@ def get_commands() -> list[AacCommand]:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/gen_protobuf/__init__.py:74]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/help_dump/__init__.py [hunk 9]
```diff
@@ -25,3 +30,18 @@ def get_commands() -> list[AacCommand]:
...
+def get_plugin() -> Plugin:
+    """
+    Returns information about the plugin.
+
+    Returns:
+        A collection of information about the plugin and what it contributes.
+    """
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/help_dump/__init__.py:45]:
```suggestion
    plugin.register_commands(get_commands())
```

File: python/src/aac/plugins/material_model/__init__.py [hunk 10]
```diff
@@ -27,9 +31,30 @@ def get_commands() -> list[AacCommand]:
...
-    return get_resource_file_contents(__package__, "material_model.yaml")
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/material_model/__init__.py:58]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/plugin.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
+"""Plugin metadata class."""
+
+from typing import Optional
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/plugin.py:3]:
```suggestion
```

File: python/src/aac/plugins/plugin.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+    def __hash__(self) -> int:
+        """Return the hash for this Plugin object."""
+        return hash(self.name)
+
+    def register_commands(self, commands: set[AacCommand]):
+        """Register the specified commands."""
+        self.contributions.register_commands(self.name, commands)
+
+    def get_commands(self) -> Optional[set[AacCommand]]:
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/plugin.py:35]:
```suggestion
    def get_commands(self) -> list[AacCommand]:
```

File: python/src/aac/plugins/plugin.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        """Get the commands registered by this plugin."""
+        return self.contributions.get_commands_by_plugin_name(self.name)
+
+    def register_validations(self, validations: set[ValidatorPlugin]):
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/plugin.py:39]:
```suggestion
    def register_validations(self, validations: list[ValidatorPlugin]):
```

File: python/src/aac/plugins/plugin.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        """Register the specified validations."""
+        self.contributions.register_validations(self.name, validations)
+
+    def get_validations(self) -> Optional[set[ValidatorPlugin]]:
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/plugin.py:43]:
```suggestion
    def get_validations(self) -> list[ValidatorPlugin]:
```

File: python/src/aac/plugins/plugin.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        """Get the validations registered by this plugin."""
+        return self.contributions.get_validations_by_plugin_name(self.name)
+
+    def register_definitions(self, definitions: set[Definition]):
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/plugin.py:47]:
```suggestion
    def register_definitions(self, definitions: list[Definition]):
```

File: python/src/aac/plugins/plugin.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        """Register the specified definitions."""
+        self.contributions.register_definitions(self.name, definitions)
+
+    def get_definitions(self) -> Optional[set[Definition]]:
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/plugin.py:51]:
```suggestion
    def get_definitions(self) -> list[Definition]:
```

File: python/src/aac/plugins/print_spec/__init__.py [hunk 17]
```diff
@@ -32,3 +35,18 @@ def get_commands() -> list[AacCommand]:
...
+def get_plugin() -> Plugin:
+    """
+    Returns information about the plugin.
+
+    Returns:
+        A collection of information about the plugin and what it contributes.
+    """
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/print_spec/__init__.py:50]:
```suggestion
    plugin.register_commands(get_commands())
```

File: python/src/aac/plugins/rest_api/__init__.py [hunk 18]
```diff
@@ -20,38 +22,40 @@ def get_commands() -> list[AacCommand]:
...
-    Return the plugins Aac definitions.
+    Returns information about the plugin.
 
     Returns:
-         string representing yaml extensions and definitions defined by the plugin
+        A collection of information about the plugin and what it contributes.
     """
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/rest_api/__init__.py:59]:
```suggestion
    plugin.register_commands(get_commands())
```

File: python/src/aac/plugins/specifications/__init__.py [hunk 19]
```diff
@@ -37,9 +41,30 @@ def get_commands() -> list[AacCommand]:
...
-    return get_resource_file_contents(__package__, "specifications.yaml")
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
+    plugin.register_definitions(set(plugin_definitions))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/specifications/__init__.py:68]:
```suggestion
    plugin.register_commands(get_commands())
    plugin.register_definitions(plugin_definitions)
```

File: python/src/aac/plugins/start_lsp/__init__.py [hunk 20]
```diff
@@ -24,22 +27,37 @@ def get_commands() -> list[AacCommand]:
...
+def get_plugin() -> Plugin:
+    """
+    Returns information about the plugin.
+
+    Returns:
+        A collection of information about the plugin and what it contributes.
+    """
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_commands(set(get_commands()))
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/start_lsp/__init__.py:61]:
```suggestion
    plugin.register_commands(get_commands())
```

File: python/src/aac/plugins/validators/defined_references/__init__.py [hunk 21]
```diff
@@ -31,3 +32,24 @@ def register_validators() -> ValidatorPlugin:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/defined_references/__init__.py:53]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/src/aac/plugins/validators/exclusive_fields/__init__.py [hunk 22]
```diff
@@ -31,3 +32,24 @@ def register_validators() -> ValidatorPlugin:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/exclusive_fields/__init__.py:53]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/src/aac/plugins/validators/reference_fields/__init__.py [hunk 23]
```diff
@@ -32,18 +34,22 @@ def register_validators() -> ValidatorPlugin:
...
-    return reference_validation and reference_validation[0].get("arguments") or []
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
 
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/reference_fields/__init__.py:53]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/src/aac/plugins/validators/reference_targets/__init__.py [hunk 24]
```diff
@@ -32,18 +34,22 @@ def register_validators() -> ValidatorPlugin:
...
-    return reference_validation and reference_validation[0].get("arguments") or []
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
 
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/reference_targets/__init__.py:53]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/src/aac/plugins/validators/required_fields/__init__.py [hunk 25]
```diff
@@ -46,4 +49,25 @@ def get_required_fields(definition: Definition) -> list[str]:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/required_fields/__init__.py:68]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/src/aac/plugins/validators/root_keys/__init__.py [hunk 26]
```diff
@@ -30,3 +32,24 @@ def register_validators() -> ValidatorPlugin:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/root_keys/__init__.py:53]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/src/aac/plugins/validators/subcomponent_type/__init__.py [hunk 27]
```diff
@@ -29,3 +32,24 @@ def register_validators() -> ValidatorPlugin:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/subcomponent_type/__init__.py:53]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/src/aac/plugins/validators/validator_implementation/__init__.py [hunk 28]
```diff
@@ -30,3 +32,24 @@ def register_validators() -> ValidatorPlugin:
...
+    """
+    plugin_definitions = parse(
+        get_plugin_aac_definitions(),
+        get_resource_file_path(*plugin_resource_file_args)
+    )
+
+    *_, plugin_name = __package__.split(".")
+    plugin = Plugin(plugin_name)
+    plugin.register_definitions(set(plugin_definitions))
+    plugin.register_validations({register_validators()})
```
Comment (Coffee2Bits) [inline python/src/aac/plugins/validators/validator_implementation/__init__.py:53]:
```suggestion
    plugin.register_definitions(plugin_definitions)
    plugin.register_validations([register_validators()])
```

File: python/tests/plugins/test_contribution_points.py [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+        self.contributions.register_commands(self.plugin_name, commands)
+        self.assertEqual(self.contributions.get_commands_by_plugin_name(self.plugin_name), commands)
+
+        validations = {create_validation(f"Test{i}", create_validation_definition(f"Validation{i}")) for i in range(self.num_items)}
+        self.contributions.register_validations(self.plugin_name, validations)
+        self.assertEqual(self.contributions.get_validations_by_plugin_name(self.plugin_name), validations)
+
+        definitions = {create_schema_definition(f"Test{i}") for i in range(self.num_items)}
+        self.contributions.register_definitions(self.plugin_name, definitions)
+        self.assertEqual(self.contributions.get_definitions_by_plugin_name(self.plugin_name), definitions)
```
Comment (Coffee2Bits) [inline python/tests/plugins/test_contribution_points.py:69]:
```suggestion
        commands = [create_command(f"Test{i}") for i in range(self.num_items)]
        self.contributions.register_commands(self.plugin_name, commands)
        self.assertListEqual(self.contributions.get_commands_by_plugin_name(self.plugin_name), commands)

        validations = [create_validation(f"Test{i}", create_validation_definition(f"Validation{i}")) for i in range(self.num_items)]
        self.contributions.register_validations(self.plugin_name, validations)
        self.assertListEqual(self.contributions.get_validations_by_plugin_name(self.plugin_name), validations)

        definitions = [create_schema_definition(f"Test{i}") for i in range(self.num_items)]
        self.contributions.register_definitions(self.plugin_name, definitions)
        self.assertListEqual(self.contributions.get_definitions_by_plugin_name(self.plugin_name), definitions)
```

File: python/tests/plugins/test_gen_json.py [hunk 30]
```diff
@@ -40,6 +48,17 @@ def test_gen_json_with_unparsable_file(self):
             result = print_json([temp_arch_file.name])
             self.assertEqual(result.status_code, PluginExecutionStatusCode.PARSER_FAILURE)
 
+    def test_get_plugin(self):
+        plugin = get_plugin()
+
+        self.assertEqual(plugin.name, "gen_json")
+
+        commands = get_commands()
+        self.assertEqual(plugin.contributions.get_commands(), set(commands))
```
Comment (Coffee2Bits) [inline python/tests/plugins/test_gen_json.py:57]:
```suggestion
        self.assertListEqual(plugin.contributions.get_commands(), commands)
```

File: python/tests/plugins/test_gen_json.py [hunk 31]
```diff
@@ -40,6 +48,17 @@ def test_gen_json_with_unparsable_file(self):
...
+
+        definitions = parse(get_plugin_aac_definitions(), get_resource_file_path(*plugin_resource_file_args))
+        self.assertEqual(plugin.contributions.get_definitions(), set(definitions))
```
Comment (Coffee2Bits) [inline python/tests/plugins/test_gen_json.py:60]:
```suggestion
        self.assertListEqual(plugin.contributions.get_definitions(), definitions)
```

**Review by Coffee2Bits — CHANGES_REQUESTED** [review]
Looks good, just some discussion/future stories and a number of suggestions to align the API to use the `list` collection like the rest of the API. 

Two more items that I couldn't really add as suggestions, but need to be addressed since you already did the heavy lifting converting the plugins:
1. Remove all the hookspecs except for `get_plugin`
2. Remove the `@hookimpl` decorator from the plugin functions except for `get_plugin`

Here is what the modified hookspecs.py file for 1. above should be:
```python
"""Defines the AaC plugin interface via Pluggy Hookspecs."""
from aac.plugins import hookspec
from aac.plugins.plugin import Plugin


@hookspec
def get_plugin() -> Plugin:
    """
    Returns information about the plugin.

    Returns:
        A collection of information about the plugin and what it contributes.
    """
```

File: python/src/aac/cli/aac_command.py [hunk 32]
```diff
@@ -44,3 +44,7 @@ class AacCommand:
     description = attrib(validator=validators.instance_of(str))
     callback = attrib(validator=validators.is_callable())
     arguments = attrib(default=Factory(list), validator=validators.instance_of(list))
+
+    def __hash__(self) -> int:
+        """Return the hash of this AacCommand."""
+        return hash(self.name)
```
[the remark this answers is not in the document]
Comment (cvchaparro) [reply python/src/aac/cli/aac_command.py:50]: I definitely think we need to have a discussion on this. I actually debated this for the `__hash__` function on `PluginContribution`. The thing that occurs to me if we incorporate argument hashes, etc is that we can end up with effectively duplicated items if we add an argument to a plugin command, etc. I created #397 discussion to brainstorm this more.

File: python/src/aac/plugins/contribution_points.py [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+from aac.plugins.contribution_type import ContributionType
+from aac.plugins.plugin_contribution import PluginContribution
+from aac.plugins.validators._validator_plugin import ValidatorPlugin
+
+
+@attrs
+class ContributionPoints:
+    """
+    A class to support contribution points which serve as hooks into AaC core for adding/customizing functionality.
+
```
Comment (cvchaparro) [inline python/src/aac/plugins/contribution_points.py:19]:
```suggestion

```

File: python/src/aac/plugins/contribution_points.py [hunk 34]
```diff
@@ -0,0 +1,144 @@
...
+from aac.plugins.contribution_type import ContributionType
+from aac.plugins.plugin_contribution import PluginContribution
+from aac.plugins.validators._validator_plugin import ValidatorPlugin
+
+
+@attrs
+class ContributionPoints:
+    """
+    A class to support contribution points which serve as hooks into AaC core for adding/customizing functionality.
+    See ContributionType for the complete list of contribution types.
```
Comment (cvchaparro) [inline python/src/aac/plugins/contribution_points.py:19]:
```suggestion

    See ContributionType for the complete list of contribution types.
```

File: python/src/aac/lang/definitions/definition.py [hunk 35]
```diff
@@ -26,6 +26,10 @@ class Definition:
     lexemes: list[Lexeme] = attrib(default=Factory(list), validator=validators.instance_of(list))
     structure: dict = attrib(default=Factory(dict), validator=validators.instance_of(dict))
 
+    def __hash__(self) -> int:
+        """Return the hash of this Definition."""
+        return hash(self.name)
```
[the remark this answers is not in the document]
Comment (cvchaparro) [reply python/src/aac/lang/definitions/definition.py:31]: See https://github.com/jondavid-black/AaC/pull/394#discussion_r928078962

File: python/tests/plugins/test_contribution_points.py [hunk 36]
```diff
@@ -0,0 +1,90 @@
...
+            create_validation(f"Test{i}", create_validation_definition(f"Validation{i}")) for i in range(self.num_items)
+        ]
+
+        self.contributions.register_commands(self.plugin_name, commands)
+        self.contributions.register_definitions(self.plugin_name, definitions)
+        self.contributions.register_validations(self.plugin_name, validations)
+
+        self.assertEqual(self.contributions.get_commands_by_plugin_name(self.plugin_name), commands)
+        self.assertEqual(self.contributions.get_definitions_by_plugin_name(self.plugin_name), definitions)
+        self.assertEqual(self.contributions.get_validations_by_plugin_name(self.plugin_name), validations)
```
Comment (cvchaparro) [inline python/tests/plugins/test_contribution_points.py:71]:
```suggestion
        self.assertCountEqual(self.contributions.get_commands_by_plugin_name(self.plugin_name), commands)
        self.assertCountEqual(self.contributions.get_definitions_by_plugin_name(self.plugin_name), definitions)
        self.assertCountEqual(self.contributions.get_validations_by_plugin_name(self.plugin_name), validations)
```

**Review by Coffee2Bits — APPROVED** [review]
Looks good!

**Pull request merged**